File: server/routes/classrooms.js
const router    = require('express').Router();
const auth      = require('../middleware/auth');
const { Classroom } = require('../models/Classroom');
const { v4: uuidv4 } = require('uuid');
const multer    = require('multer');
const Document  = require('../models/Document');
const { chunkText } = require('../services/ai');

const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 30 * 1024 * 1024 } });

// POST /api/classrooms — create classroom (teacher only)
router.post('/', auth, async (req, res) => {
  try {
    if (!['teacher', 'admin'].includes(req.user.role))
      return res.status(403).json({ message: 'Only teachers can create classrooms' });

    const { name, subject, description, schedule } = req.body;
    const joinCode = uuidv4().slice(0, 8).toUpperCase();
    const room = await Classroom.create({ teacher: req.user._id, name, subject, description, schedule, joinCode });
    res.status(201).json(room);
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// POST /api/classrooms/join — student join
router.post('/join', auth, async (req, res) => {
  try {
    const { join_code } = req.body;
    const room = await Classroom.findOne({ joinCode: join_code });
    if (!room) return res.status(404).json({ message: 'Classroom not found' });
    if (room.teacher.toString() === req.user._id.toString())
      return res.status(400).json({ message: 'Teachers cannot enroll in their own class' });
    if (!room.enrollments.includes(req.user._id)) {
      room.enrollments.push(req.user._id);
      await room.save();
    }
    res.json(room);
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// GET /api/classrooms — list for current user
router.get('/', auth, async (req, res) => {
  try {
    const uid = req.user._id;
    const normalize = c => ({
      id: c._id, _id: c._id,
      name: c.name, subject: c.subject,
      description: c.description, schedule: c.schedule,
      join_code: c.joinCode, joinCode: c.joinCode,
      is_active: c.isActive, teacher: c.teacher,
      createdAt: c.createdAt
    });

    if (['teacher', 'admin'].includes(req.user.role)) {
      const teaching = await Classroom.find({ teacher: uid });
      const enrolled = await Classroom.find({ enrollments: uid });
      return res.json({ teaching: teaching.map(normalize), enrolled: enrolled.map(normalize) });
    }
    const enrolled = await Classroom.find({ enrollments: uid });
    res.json({ teaching: [], enrolled: enrolled.map(normalize) });
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// GET /api/classrooms/:id — detail with lectures + students + materials
router.get('/:id', auth, async (req, res) => {
  try {
    const { Lecture } = require('../models/Classroom');
    const User = require('../models/User');

    const room = await Classroom.findById(req.params.id).populate('teacher', 'email');
    if (!room) return res.status(404).json({ message: 'Classroom not found' });

    // Access check
    const isTeacher  = room.teacher._id.toString() === req.user._id.toString() || req.user.role === 'admin';
    const isEnrolled = room.enrollments.map(e => e.toString()).includes(req.user._id.toString());
    if (!isTeacher && !isEnrolled) return res.status(403).json({ message: 'Not enrolled' });

    const lectures = await Lecture.find({ classroom: req.params.id }).sort({ createdAt: -1 })
      .select('title isLive createdAt endedAt detectedTopics summary autoQuiz');

    const lectureList = lectures.map(l => ({
      id: l._id, title: l.title, is_live: l.isLive,
      started_at: l.createdAt, ended_at: l.endedAt,
      detected_topics: l.detectedTopics,
      has_summary: !!l.summary, has_quiz: Array.isArray(l.autoQuiz) && l.autoQuiz.length > 0
    }));

    const students = await User.find({ _id: { $in: room.enrollments } }).select('email createdAt');

    res.json({
      id: room._id, name: room.name, subject: room.subject,
      description: room.description, schedule: room.schedule,
      join_code: room.joinCode, teacher_id: room.teacher._id,
      is_active: room.isActive, student_count: room.enrollments.length,
      lectures: lectureList, materials: room.materials,
      students: students.map(s => ({ id: s._id, email: s.email }))
    });
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// DELETE /api/classrooms/:id
router.delete('/:id', auth, async (req, res) => {
  const room = await Classroom.findById(req.params.id);
  if (!room) return res.status(404).json({ message: 'Not found' });
  if (room.teacher.toString() !== req.user._id.toString() && req.user.role !== 'admin')
    return res.status(403).json({ message: 'Not authorized' });
  await room.deleteOne();
  res.json({ message: 'Classroom deleted' });
});

// POST /api/classrooms/:id/materials
router.post('/:id/materials', auth, async (req, res) => {
  try {
    const room = await Classroom.findById(req.params.id);
    if (!room) return res.status(404).json({ message: 'Not found' });
    if (room.teacher.toString() !== req.user._id.toString() && req.user.role !== 'admin')
      return res.status(403).json({ message: 'Only teacher can add materials' });

    const { title, content, material_type } = req.body;
    room.materials.push({ title, content, materialType: material_type || 'note' });
    await room.save();
    res.json(room.materials[room.materials.length - 1]);
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// DELETE /api/classrooms/:id/materials/:matId
router.delete('/:id/materials/:matId', auth, async (req, res) => {
  try {
    const room = await Classroom.findById(req.params.id);
    if (!room) return res.status(404).json({ message: 'Not found' });
    room.materials = room.materials.filter(m => m._id.toString() !== req.params.matId);
    await room.save();
    res.json({ message: 'Material deleted' });
  } catch (err) { res.status(500).json({ message: err.message }); }
});

// POST /api/classrooms/:id/materials/upload — teacher uploads a file as class material
// It also creates a Document in ALL enrolled students' knowledge bases
router.post('/:id/materials/upload', auth, upload.single('file'), async (req, res) => {
  try {
    const room = await Classroom.findById(req.params.id).populate('enrollments', '_id');
    if (!room) return res.status(404).json({ message: 'Not found' });
    if (room.teacher.toString() !== req.user._id.toString() && req.user.role !== 'admin')
      return res.status(403).json({ message: 'Only the teacher can upload class materials' });
    if (!req.file) return res.status(400).json({ message: 'No file provided' });

    const title    = req.file.originalname;
    const ext      = title.split('.').pop().toLowerCase();
    const sourceType = ext === 'pdf' ? 'pdf' : 'txt';

    // Extract text
    let text = '';
    if (ext === 'pdf') {
      try {
        const pdfParse = require('pdf-parse');
        const data = await pdfParse(req.file.buffer);
        text = data.text || '';
      } catch(e) { text = req.file.buffer.toString('utf-8'); }
    } else {
      text = req.file.buffer.toString('utf-8');
    }

    const chunks = text.trim().length > 10 ? chunkText(text) : [];
    const classLabel = `[${room.name}] ${title}`;

    // Save as material in classroom (text content for inline view)
    room.materials.push({ title: classLabel, content: text.slice(0, 2000), materialType: ext });
    await room.save();
    const newMaterial = room.materials[room.materials.length - 1];

    // Push Document to all enrolled students' knowledge bases
    const studentIds = room.enrollments.map(e => e._id || e);
    const docInserts = studentIds.map(uid => ({
      user: uid, title: classLabel, content: text, sourceType, chunks
    }));
    if (docInserts.length > 0) {
      await Document.insertMany(docInserts);
    }

    res.status(201).json({ material: newMaterial, pushed_to: studentIds.length + ' students' });
  } catch (err) { res.status(500).json({ message: err.message }); }
});

module.exports = router;

File: backend/app/api/endpoints/classrooms.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.api.deps import get_current_user
from app.db.models import User, Classroom, ClassEnrollment, Lecture, ClassMaterial
from app.schemas import ClassroomCreate, ClassroomJoin, ClassroomResponse, MaterialCreate, MaterialResponse
import uuid

router = APIRouter()


# ═══════════════════════════════════════
# CLASSROOM CRUD
# ═══════════════════════════════════════

@router.post("/", response_model=ClassroomResponse)
def create_classroom(req: ClassroomCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Teacher only: Create a new classroom with subject and schedule."""
    if current_user.role not in ("teacher", "admin"):
        raise HTTPException(status_code=403, detail="Only teachers can create classrooms")

    join_code = str(uuid.uuid4())[:8].upper()

    classroom = Classroom(
        teacher_id=current_user.id,
        name=req.name,
        subject=req.subject,
        description=req.description,
        schedule=req.schedule,
        join_code=join_code
    )
    db.add(classroom)
    db.commit()
    db.refresh(classroom)
    return classroom


@router.post("/join", response_model=ClassroomResponse)
def join_classroom(req: ClassroomJoin, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Student: Join a classroom via join code."""
    classroom = db.query(Classroom).filter(Classroom.join_code == req.join_code).first()
    if not classroom:
        raise HTTPException(status_code=404, detail="Classroom not found with this code")

    enrolled = db.query(ClassEnrollment).filter(
        ClassEnrollment.classroom_id == classroom.id,
        ClassEnrollment.student_id == current_user.id
    ).first()
    if enrolled:
        return classroom

    if current_user.id == classroom.teacher_id:
        raise HTTPException(status_code=400, detail="Teachers cannot enroll in their own classroom")

    enrollment = ClassEnrollment(classroom_id=classroom.id, student_id=current_user.id)
    db.add(enrollment)
    db.commit()
    return classroom


@router.get("/")
def get_classrooms(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Get classrooms — teachers see taught classes, students see enrolled."""
    if current_user.role in ("teacher", "admin"):
        teaching = db.query(Classroom).filter(Classroom.teacher_id == current_user.id).all()
        # Teachers can also be enrolled in other classes
        enrollments = db.query(ClassEnrollment).filter(ClassEnrollment.student_id == current_user.id).all()
        class_ids = [e.classroom_id for e in enrollments]
        enrolled = db.query(Classroom).filter(Classroom.id.in_(class_ids)).all() if class_ids else []
        return {"teaching": teaching, "enrolled": enrolled}
    else:
        enrollments = db.query(ClassEnrollment).filter(ClassEnrollment.student_id == current_user.id).all()
        class_ids = [e.classroom_id for e in enrollments]
        enrolled = db.query(Classroom).filter(Classroom.id.in_(class_ids)).all() if class_ids else []
        return {"teaching": [], "enrolled": enrolled}


@router.get("/{classroom_id}")
def get_classroom_details(classroom_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Get classroom details including lectures, materials, students."""
    classroom = db.query(Classroom).filter(Classroom.id == classroom_id).first()
    if not classroom:
        raise HTTPException(status_code=404, detail="Classroom not found")

    # Check access
    if classroom.teacher_id != current_user.id and current_user.role != "admin":
        enrolled = db.query(ClassEnrollment).filter(
            ClassEnrollment.classroom_id == classroom.id,
            ClassEnrollment.student_id == current_user.id
        ).first()
        if not enrolled:
            raise HTTPException(status_code=403, detail="Not enrolled in this classroom")

    lectures = db.query(Lecture).filter(Lecture.classroom_id == classroom.id).order_by(Lecture.started_at.desc()).all()
    materials = db.query(ClassMaterial).filter(ClassMaterial.classroom_id == classroom.id).order_by(ClassMaterial.created_at.desc()).all()
    enrollments = db.query(ClassEnrollment).filter(ClassEnrollment.classroom_id == classroom.id).all()

    students = []
    for e in enrollments:
        student = db.query(User).filter(User.id == e.student_id).first()
        if student:
            students.append({"id": student.id, "email": student.email, "enrolled_at": e.enrolled_at.isoformat()})

    lecture_list = []
    for l in lectures:
        lecture_list.append({
            "id": l.id,
            "title": l.title,
            "is_live": l.is_live,
            "started_at": l.started_at.isoformat() if l.started_at else None,
            "ended_at": l.ended_at.isoformat() if l.ended_at else None,
            "detected_topics": l.detected_topics,
            "has_summary": bool(l.summary),
            "has_quiz": l.auto_quiz != "[]"
        })

    material_list = []
    for m in materials:
        material_list.append({
            "id": m.id, "title": m.title, "content": m.content,
            "material_type": m.material_type, "created_at": m.created_at.isoformat()
        })

    return {
        "id": classroom.id,
        "name": classroom.name,
        "subject": classroom.subject,
        "description": classroom.description,
        "schedule": classroom.schedule,
        "join_code": classroom.join_code,
        "teacher_id": classroom.teacher_id,
        "is_active": classroom.is_active,
        "lectures": lecture_list,
        "materials": material_list,
        "students": students,
        "student_count": len(students)
    }


@router.delete("/{classroom_id}")
def delete_classroom(classroom_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Teacher only: Delete a classroom."""
    classroom = db.query(Classroom).filter(Classroom.id == classroom_id).first()
    if not classroom:
        raise HTTPException(status_code=404, detail="Classroom not found")
    if classroom.teacher_id != current_user.id and current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Not authorized")

    db.delete(classroom)
    db.commit()
    return {"message": "Classroom deleted"}


# ═══════════════════════════════════════
# MATERIALS (pre-class uploads)
# ═══════════════════════════════════════

@router.post("/{classroom_id}/materials", response_model=MaterialResponse)
def add_material(classroom_id: int, req: MaterialCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Teacher only: Add pre-class material."""
    classroom = db.query(Classroom).filter(Classroom.id == classroom_id).first()
    if not classroom:
        raise HTTPException(status_code=404, detail="Classroom not found")
    if classroom.teacher_id != current_user.id and current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Only the teacher can add materials")

    material = ClassMaterial(
        classroom_id=classroom.id,
        title=req.title,
        content=req.content,
        material_type=req.material_type
    )
    db.add(material)
    db.commit()
    db.refresh(material)
    return material


@router.delete("/{classroom_id}/materials/{material_id}")
def delete_material(classroom_id: int, material_id: int, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    """Teacher only: Delete a pre-class material."""
    material = db.query(ClassMaterial).filter(ClassMaterial.id == material_id, ClassMaterial.classroom_id == classroom_id).first()
    if not material:
        raise HTTPException(status_code=404, detail="Material not found")

    classroom = db.query(Classroom).filter(Classroom.id == classroom_id).first()
    if classroom.teacher_id != current_user.id and current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Not authorized")

    db.delete(material)
    db.commit()
    return {"message": "Material deleted"}
